tests/unit/test_query_graph.cpp

```diff
@@ -336,3 +337,97 @@ TEST_F(QueryGraphTest, QueryGraphConnectedComponents) {
 	array_free(connected_components);
 }
 
+TEST_F(QueryGraphTest, QueryGraphExtractSubGraph) {
+	//--------------------------------------------------------------------------
+	// Construct graph
+	//--------------------------------------------------------------------------
+
+	// (A)->(B)->(C)->(D)
+	const char *relation = "R";
+
+	QGNode *A = QGNode_New("A");
+	QGNode *B = QGNode_New("B");
+	QGNode *C = QGNode_New("C");
+	QGNode *D = QGNode_New("D");
+
+	QGEdge *AB = QGEdge_New(A, B, relation, "AB");
+	QGEdge *BC = QGEdge_New(B, C, relation, "BC");
+	QGEdge *CD = QGEdge_New(C, D, relation, "CD");
+
+	uint node_cap = 4;
+	uint edge_cap = 3;
+	QueryGraph *qg = QueryGraph_New(node_cap, edge_cap);
+	QueryGraph_AddNode(qg, A);
+	QueryGraph_AddNode(qg, B);
+	QueryGraph_AddNode(qg, C);
+	QueryGraph_AddNode(qg, D);
+
+	QueryGraph_ConnectNodes(qg, A, B, AB);
+	QueryGraph_ConnectNodes(qg, B, C, BC);
+	QueryGraph_ConnectNodes(qg, C, A, CD);
+
+	//--------------------------------------------------------------------------
+	// Extract portions of the original query graph
+	//--------------------------------------------------------------------------
+
+	const char *query = "MATCH (A)-[AB]->(B), (B)-[BC]->(C) MATCH (C)-[CD]->(D) MATCH (D)-[DE]->(E)RETURN D";
+	cypher_parse_result_t *parse_result = cypher_parse(query, NULL, NULL, CYPHER_PARSE_ONLY_STATEMENTS);
+	AST *ast = AST_Build(parse_result);
+	ast->referenced_entities = raxNew();
+
+	const cypher_astnode_t **match_clauses = AST_GetClauses(ast, CYPHER_AST_MATCH);
+	const cypher_astnode_t *patterns[3];
+
+	// Extract patterns, one per MATCH clause
+	patterns[0] = cypher_ast_match_get_pattern(match_clauses[0]);
+	patterns[1] = cypher_ast_match_get_pattern(match_clauses[1]);
+	patterns[2] = cypher_ast_match_get_pattern(match_clauses[2]);
+
+	// Empty sub graph, as the number of patterns specified is 0.
+	QueryGraph *sub = QueryGraph_ExtractPatterns(qg, patterns, 0);
+
+	// Validation, expecting an empty query graph.
+	ASSERT_EQ(QueryGraph_NodeCount(sub), 0);
+	ASSERT_EQ(QueryGraph_EdgeCount(sub), 0);
+	QueryGraph_Free(sub);
+
+	// Single pattern, 2 paths, sub graph: a->b->c
+	sub = QueryGraph_ExtractPatterns(qg, patterns, 1);
```

Comment:
I get the error:
```
assert(ctx->gc) failed: /home/tower-linux/dev/redis/graph/src/query_ctx.c line 131
test_query_graph.run: /home/tower-linux/dev/redis/graph/src/query_ctx.c:131: QueryCtx_GetGraphCtx: Assertion `false' failed.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff6a838b1 in __GI_abort () at abort.c:79
#2  0x00007ffff6a7342a in __assert_fail_base (fmt=0x7ffff6bfaa38 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n",
    assertion=assertion@entry=0x555555c30b52 "false", file=file@entry=0x555555c30b20 "/home/tower-linux/dev/redis/graph/src/query_ctx.c",
    line=line@entry=131, function=function@entry=0x555555c30d30 <__PRETTY_FUNCTION__.16823> "QueryCtx_GetGraphCtx") at assert.c:92
#3  0x00007ffff6a734a2 in __GI___assert_fail (assertion=0x555555c30b52 "false", file=0x555555c30b20 "/home/tower-linux/dev/redis/graph/src/query_ctx.c",
    line=131, function=0x555555c30d30 <__PRETTY_FUNCTION__.16823> "QueryCtx_GetGraphCtx") at assert.c:101
#4  0x000055555560179a in QueryCtx_GetGraphCtx () at /home/tower-linux/dev/redis/graph/src/query_ctx.c:131
#5  0x00005555556e8310 in _QueryGraphAddEdge (qg=0x5555561d1640, ast_entity=0x5555561d0890, src=0x5555561d17f0, dest=0x5555561d1830)
    at /home/tower-linux/dev/redis/graph/src/graph/query_graph.c:63
#6  0x00005555556e87ea in _QueryGraph_ExtractEdge (qg=0x5555561cfc40, graph=0x5555561d1640, left=0x5555561d17f0, right=0x5555561d1830, ast=0x5555561cedd0,
    ast_edge=0x5555561d0890) at /home/tower-linux/dev/redis/graph/src/graph/query_graph.c:157
#7  0x00005555556e8978 in _QueryGraph_ExtractPath (qg=0x5555561cfc40, graph=0x5555561d1640, path=0x5555561d09e0)
    at /home/tower-linux/dev/redis/graph/src/graph/query_graph.c:193
#8  0x00005555556e8e71 in QueryGraph_ExtractPatterns (qg=0x5555561cfc40, patterns=0x7fffffffcd40, n=1)
    at /home/tower-linux/dev/redis/graph/src/graph/query_graph.c:284
#9  0x00005555555abb0a in QueryGraphTest_QueryGraphExtractSubGraph_Test::TestBody (this=0x5555561cfb40) at test_query_graph.cpp:395
```
This file might need to mock more of a QueryCtx (though I'm not sure why this is just a local error).